internal/fs/kernel_list_cache_test.go

```diff
@@ -0,0 +1,353 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// A collection of tests which tests the kernel-list-cache feature, in which
+// directory listing 2nd time is served from kernel page-cache unless not invalidated.
+// Base of all the tests: how to detect if directory listing is served from page-cache
+// or from GCSFuse?
+// (a) GCSFuse file-system ensures different content, when listing happens on the same directory.
+// (b) If two consecutive directory listing for the same directory are same, that means
+//     2nd listing is served from kernel-page-cache.
+// (c) If not then, both 1st and 2nd listing are served from GCSFuse filesystem.
+
+package fs_test
+
+import (
+	"os"
+	"path"
+	"testing"
+	"time"
+
+	"github.com/googlecloudplatform/gcsfuse/v2/internal/config"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/suite"
+)
+
+const (
+	kernelListCacheTtlSeconds = 1000
+)
+
+type KernelListCacheTestCommon struct {
+	suite.Suite
+	fsTest
+}
+
+func (t *KernelListCacheTestCommon) SetupSuite() {
+	t.serverCfg.ImplicitDirectories = true
+	t.serverCfg.MountConfig = &config.MountConfig{
+		ListConfig: config.ListConfig{
+			KernelListCacheTtlSeconds: kernelListCacheTtlSeconds,
+		}}
+	t.serverCfg.RenameDirLimit = 10
+	t.fsTest.SetUpTestSuite()
+}
+
+func (t *KernelListCacheTestCommon) SetupTest() {
+	t.createFilesAndDirStructureInBucket()
+	cacheClock.SetTime(time.Date(2015, 4, 5, 2, 15, 0, 0, time.Local))
+}
+
+func (t *KernelListCacheTestCommon) TearDownTest() {
+	t.fsTest.TearDown()
+}
+
+func (t *KernelListCacheTestCommon) TearDownSuite() {
+	t.fsTest.TearDownTestSuite()
+}
+
+func TestKernelListCacheTestSuite(t *testing.T) {
+	suite.Run(t, new(KernelListCacheTestCommon))
+}
+
+// createFilesAndDirStructureInBucket creates the following files and directory
+// structure.
+// bucket
+//
+//	explicitDir/
+//	explicitDir/file1.txt
+//	explicitDir/file2.txt
+//	implicitDir/file1.txt
+//	implicitDir/file2.txt
+func (t *KernelListCacheTestCommon) createFilesAndDirStructureInBucket() {
+	assert.Nil(t.T(), t.createObjects(map[string]string{
+		"explicitDir/":          "",
+		"explicitDir/file1.txt": "12345",
+		"explicitDir/file2.txt": "6789101112",
+		"implicitDir/file1.txt": "-1234556789",
+		"implicitDir/file2.txt": "kdfkdj9",
+	}))
+}
+
+func (t *KernelListCacheTestCommon) deleteObjectOrFail(objectName string) {
+	assert.Nil(t.T(), t.deleteObject(objectName))
+}
+
+type KernelListCacheTestWithPositiveTtl struct {
+	suite.Suite
+	fsTest
+	KernelListCacheTestCommon
+}
+
+func (t *KernelListCacheTestWithPositiveTtl) SetupSuite() {
+	t.serverCfg.ImplicitDirectories = true
```

Comment:
Just to get more info: will it work with ImplicitDirectories = false?

Reply:
+1, in addition to this, can we add some tests for implicit dir cache validation too?

Reply:
Yes, this will work for implicitDir = false.

So, whatever GCSFuse will return to kernel as part of readdir() response, that will be cached in the page-cache. Validated this.

Added the cache hit/miss case for implicit-dir.